### PR TITLE
feat(front-end): Working search box and results

### PR DESCRIPTION
**Pull request by nchapman**
Working search results are here! Lots of new views. The header is now a little stack of specialized views. The visit presenter has been reworked as a user page presenter. Tried to share a reasonable amount of code between visits and search. I avoided making many scss changes since @johngruen is working on that in a separate branch.

@6a68 @pdehaan r?

### CODE REVIEW

File: app/scripts/templates/global_header/search_box.html [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<form>
+  <input class='query' type="text" placeholder="Search" />
```
Comment (pdehaan) [inline app/scripts/templates/global_header/search_box.html]: `class="query"` (double quotes for consistency)

File: app/scripts/templates/user_pages/item.html [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<a class="image {{imagePosition}}" href="{{url}}">
+<a class="image {{imagePosition}}" href="{{_getUrl}}">
```
Comment (pdehaan) [inline app/scripts/templates/user_pages/item.html]: This looks weird, why the underscore in `{{_getUrl}}`? The underscore makes me think its a private variable or something.

Comment (nchapman) [inline app/scripts/templates/user_pages/item.html]: Yeah it's on purpose. It's a hack until we fix the data. I want it to look wrong so we'll fix it :smile:

File: app/scripts/templates/user_pages/item.html [hunk 3]
```diff
@@ -6,9 +6,9 @@
     <img class="favicon" src="{{faviconUrl}}" />
```
Comment (pdehaan) [inline app/scripts/templates/user_pages/item.html]: Not sure if we want to add an `alt` or `title` tag here like a good citizen. Not sure if we're always guaranteed to show the `{{extractedProviderName}}` in the .medium or .small views.

File: app/scripts/templates/user_pages/item.html (hunk 2 above)
Comment (nchapman) [inline app/scripts/templates/user_pages/item.html]: See: https://github.com/mozilla/chronicle/pull/265/files#diff-1eb2942a07dc6b7ba688257f29b1313aR143

File: app/scripts/collections/search_results.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (pdehaan) [inline app/scripts/collections/search_results.js]: I think we discussed this earlier, but it's mildly weird to have underscores in filenames in _/app/*_ but dashes in _/server/*_.

For example: _/app/scripts/collections/**search_results.js**_ vs _/server/work-queue/jobs/**extract-page.js**_.

File: app/scripts/collections/search_results.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  'backbone',
+  'models/user_page'
+], function (_, Backbone, UserPage) {
+  'use strict';
+
+  var SearchResults = Backbone.Collection.extend({
+    model: UserPage,
+    url: '/v1/search',
+
+    parse: function (response, xhr) {
```
Comment (pdehaan) [inline app/scripts/collections/search_results.js]: `xhr` is defined but never used.

File: app/scripts/views/global_header/search_box.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define([
+  'backbone',
+  'views/base',
+  'stache!templates/global_header/search_box'
+], function (Backbone, BaseView, SearchBoxTemplate, router) {
```
Comment (pdehaan) [inline app/scripts/views/global_header/search_box.js]: `router` is defined but never used.

File: app/scripts/templates/search/index.html [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<h3 class="number-of-results">
+  {{#onlyOneResult}}
+    {{numberOfResults}} result
+  {{/onlyOneResult}}
+  {{^onlyOneResult}}
+    {{numberOfResults}} results
+  {{/onlyOneResult}}
```
Comment (pdehaan) [inline app/scripts/templates/search/index.html]: Probably not the best place for my kvetching, but we need a better state for "0 results" as it looks pretty barren.

Or tell me to shut up and I'll file a new bug and we can unleash the UX wonder-twins on it.

**Comment by pdehaan** [comment]
Tested locally and everything seems kosher, but I'll let @6a68 take a look.
Apart from the 0 search results emptyness, I like it! I'll be nice to see this land in the dev server so I can play more with my real sites versus my local dummy data.

Actually, I *am* seeing some slight funk when I look at my dummy data and see a reference to "Klee" (on our Tumblr blog) but I don't see any results when I search for "klee" or "Klee".

**Comment by 6a68** [comment]
@pdehaan If you aren't seeing search results, that's more a problem with our elasticsearch query than the front-end code here. I'm personally not that bothered by "0 results", as it communicates the result of the user's action in a grammatically correct, if minimal, way :-)

File: app/scripts/collections/search_results.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+      if (response && response.results && response.results.hits) {
```
Comment (6a68) [inline app/scripts/collections/search_results.js]: ugh, the search API JSON should not force you to do stuff like this. added it to the list of nits in #264

File: app/scripts/views/global_header/search_box.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    events: {
+      'submit form': 'search'
+    },
+
+    search: function (event) {
+      event.preventDefault();
+
+      var encodedQuery = window.encodeURIComponent(this.$('.query').val());
+
+      Backbone.history.navigate('search/' + encodedQuery, { trigger: true });
```
Comment (6a68) [inline app/scripts/views/global_header/search_box.js]: we probably want to eventually do something nice if the user empties out the search box and hits enter, but this is fine for now

**Comment by 6a68** [comment]
This looks great, @nchapman!

Looks like it needs a rebase. In the interest of moving things along, I'm going to rebase this for you and merge it!

:beers: and :cake: for everyone! :fireworks:

File: app/scripts/collections/search_results.js (hunk 8 above)
Comment (pdehaan) [inline app/scripts/collections/search_results.js]: We could switch to CoffeeScript, and just do something like `if response?.results?.hits`
(ducks and hides)

Comment (kitcambridge) [inline app/scripts/collections/search_results.js]: @pdehaan Or an [`Option`](https://github.com/fantasyland/fantasy-options) monad. :stuck_out_tongue_winking_eye:

**Pull request merged**